quickwit/quickwit-ingest/src/ingest_v2/routing_table.rs

```diff
@@ -0,0 +1,835 @@
+                    return Some(shard);
+                }
+            }
+        }
+        None
+    }
+
+    /// Inserts the open shards the routing table is not aware of.
+    fn insert_open_shards(
+        &mut self,
+        self_node_id: &NodeId,
+        leader_id: &NodeId,
+        index_uid: &IndexUid,
+        shard_ids: &[ShardId],
+    ) {
+        match self.index_uid.cmp(index_uid) {
+            // If we receive an update for a new incarnation of the index, then we clear the entry
+            // and insert all the shards.
+            std::cmp::Ordering::Less => {
+                self.index_uid = index_uid.clone();
+                self.clear_shards();
+            }
+            // If we receive an update for a previous incarnation of the index, then we ignore it.
+            std::cmp::Ordering::Greater => {
+                return;
+            }
+            std::cmp::Ordering::Equal => {}
+        };
+        let target_shards = if self_node_id == leader_id {
+            &mut self.local_shards
+        } else {
+            &mut self.remote_shards
+        };
+        let target_shards_len = target_shards.len();
+        let mut num_inserted_shards = 0;
+
+        for &shard_id in shard_ids {
+            // If we can't find the shard, then we insert it.
+            if target_shards[..target_shards_len]
+                .binary_search_by_key(&shard_id, |shard| shard.shard_id)
+                .is_err()
+            {
+                target_shards.push(RoutingEntry {
+                    index_uid: self.index_uid.clone(),
+                    source_id: self.source_id.clone(),
+                    shard_id,
+                    shard_state: ShardState::Open,
+                    leader_id: leader_id.clone(),
+                });
+                num_inserted_shards += 1;
+            }
+        }
+        if num_inserted_shards > 0 {
+            target_shards.sort_unstable_by_key(|shard| shard.shard_id);
+
+            info!(
+                index_id=%self.index_uid.index_id(),
+                source_id=%self.source_id,
+                "inserted {num_inserted_shards} shards into routing table"
+            );
+        }
+    }
+
+    /// Clears local and remote shards.
+    fn clear_shards(&mut self) {
+        self.local_shards.clear();
+        self.local_round_robin_idx = AtomicUsize::default();
+        self.remote_shards.clear();
+        self.remote_round_robin_idx = AtomicUsize::default();
+    }
+
+    /// Closes the shards identified by their shard IDs.
+    fn close_shards(&mut self, index_uid: &IndexUid, shard_ids: &[ShardId]) {
+        // If the shard table was just recently updated with shards for a new index UID, then we can
+        // safely discard this request.
+        if self.index_uid != *index_uid {
+            return;
+        }
+        for shards in [&mut self.local_shards, &mut self.remote_shards] {
```

Comment:
should we just entirely remove the entry if we are in the remote shards?
If not, when will be removed?

I forgot the other PR. Do we cleanup this metadata for local shard on EOF truncate?

Reply:
I'm currently adding the removal of deleted shards from the routing table to this PR.